Use the CommandInterface for constants instead of Command class.

ADmad opened the pull request:
Refs #13930.

The use of `VersionCommand` and `HelpCommand` needs to be addressed. We can either move then under `Console` or add check to use them only if the class exists.

ADmad committed:
Use the CommandInterface for constants instead of Command class.

othercorey commented:
Maybe not for this PR, but would it make sense to move Command.php to Commands/?  It stands out in the package and would be unused.

ADmad committed:
Rename methods.

Since shells will be removed in future it's better to use "command" in method names.

ADmad commented:
I renamed couple of protected methods. Since shells will be removed in future it's better to use "command" in method names unless the method deals with shells only.

ADmad committed:
Move HelpCommand under Console.

This it's available in the standalone package.

ADmad commented:
@markstory Should I move `Command` class under `Command` namespace as @othercorey suggested and add class alias for it's existing location?

othercorey commented:
Are you fixing the dependency on Core\Configuration as well with this?

ADmad commented:
> Are you fixing the dependency on Core\Configuration as well with this?

`composer.json` of the console package does have `cake/core` as a requirement which contains the `Configure` class. 

The problem though is that `Configure::version()` uses `CORE_PATH` constant to generate path to core's `config/config.php` which too uses `CORE_PATH` constant. We could replace the use of constants with relative paths.

othercorey commented:
I am not sure the usefulness of cake's version number in an independent console though.

ADmad commented:
I don't have a strong opinion on it. So I have fine we doing what I suggested or moving back the `VersionCommand` under `Command` and loading in conditionally.

othercorey commented:
I think a common version command would help.  I'm sure we can come up wit something in a PR after this.  I'll let mark decide though.

ADmad committed:
Move Console/Command under Command/Command.

The Command class has dependency on ORM package and hence having it
under standalone console package doesn't seem right.

ADmad committed:
Avoid using CORE_PATH constant to read VERSION.txt file.

That constant is only available in apps using the full framework.

ADmad commented:
So I just went ahead and moved the `Command` class and updated code to fetch CakePHP version not rely on the `CORE_PATH` constant.

ADmad commented:
> updated code to fetch CakePHP version not rely on the CORE_PATH constant.

Still doesn't fix the problem as neither `config/config.php` nor `src/VERSION.txt` are going to be available in the standalone console package.

markstory commented on src/Console/Command.php:
```diff
@@ -1,63 +1,10 @@
+    'Cake\Console\Command'
+);
+deprecationWarning(
+    'Use Cake\Command\Command instead of Cake\Console\Command.'
```
This is going to be a noisy deprecation. Can we leave this as a silent alias for the time being? This adds toil to upgrading with little value, as we learned with the ConsoleErrorHandler move uses get tripped up by these.

ADmad replied:
Yup I was thinking that too. Will remove the warning.

markstory replied:
> Should I move Command class under Command namespace as @othercorey suggested and add class alias for it's existing location?

The `Command` class being in the package is only a problem if people try to use it. Could we solve that problem with documentation instead of adding deprecation warnings that impact applications and plugins for a standalone package that may not have widespread adoption.

othercorey replied:
So, as a new user, I followed the documentation on the README as that was the first and only thing I saw.  There is no mention of the Command class there.  However, if you click on the links a the bottom for other things, you can find the Command Objects documentation which starts with explaining using the Command class.

It's awkward it exists but isn't usable, but this is just FYI.  I'm not making an argument either way.

markstory commented on tests/TestCase/Console/CommandCollectionTest.php:
```diff
@@ -272,7 +270,6 @@ public function testAutoDiscoverCore()
         // These have to be strings as ::class uses the local namespace.
         $this->assertSame(RoutesCommand::class, $collection->get('routes'));
         $this->assertSame(SampleShell::class, $collection->get('sample'));
-        $this->assertSame(VersionCommand::class, $collection->get('version'));
```
Shouldn't the version command still be present?

ADmad replied:
The test is for auto discovered commands. Auto discovery only looks under src/Command so it's not present. 

Anyway I'll move it back as getting version is not possible in standalone packages.

markstory reviewed (commented):
I am not concerned about the VersionCommand being available in standalone usage. It could stay in it's current namespace and not be added to the command collection conditionally.

It would be good to have the HelpCommand though, as it is tedious to reimplement in each application.

ADmad committed:
Ensure VersionCommand class is available.

It won't be available when using just console package.

ADmad commented:
- Moved back `VersionCommand` under `Command`.
- Removed the warning triggered in `src/Console/Command.php`.

ADmad commented:
The checks job failure is due to pslam crashing. I'll open a ticket for it.

othercorey commented:
Looks ok to me.

markstory commented on config/config.php:
```diff
@@ -14,7 +14,7 @@
  * @license       https://opensource.org/licenses/mit-license.php MIT License
  */
-$versionFile = file(CORE_PATH . 'VERSION.txt');
+$versionFile = file(dirname(__DIR__) . '/VERSION.txt');
```
I was thinking about this problem some more, and will put up a pull request to make `Configure::version()` better.

othercorey replied:
It would be kind of nice to have a build in version command that you could just configure ... but it doesn't make or break Console imo.

markstory commented on src/Console/CommandCollection.php:
```diff
@@ -203,6 +203,9 @@ protected function resolveNames(array $input): array
         }
 
+        // This entry is created for the `Command` class, which we don't need.
+        unset($out['']);
```
Could we exclude the base class in CommandScanner so it doesn't show up in other places as well?

ADmad replied:
Done

ADmad commented:
Anything else to be done here?
